File: rust/lib/src/prime.rs
use std::intrinsics::{ceilf32, logf32};

fn main_helper(n: usize) -> (Vec<u32>, Vec<bool>) {
    let mut sieve: Vec<bool> = Vec::with_capacity(n + 1);
    let mut primes: Vec<u32> = Vec::new();
    for _ in 0..=n {
        sieve.push(true);
    }
    for x in 2..=((n as f64).sqrt() as usize) {
        if sieve[x] {
            for y in (x * x..=n).step_by(x) {
                sieve[y] = false;
            }
            primes.push(x as u32);
        }
    }
    (primes, sieve)
}

pub fn make_sieve(n: usize) -> Vec<bool> {
    let (_, sieve) = main_helper(n);
    sieve
}

pub fn get_sieve_and_primes(n: usize) -> (Vec<u32>, Vec<bool>) {
    let (mut primes, sieve) = main_helper(n);
    for x in ((n as f64).sqrt().ceil() as usize)..=n {
        if sieve[x] {
            primes.push(x as u32);
        }
    }
    (primes, sieve)
}

pub fn get_primes_upto(n: usize) -> Vec<u32> {
    let (primes, _) = get_sieve_and_primes(n);
    primes
}

pub fn get_nth_prime(n: usize) -> u32 {
    unsafe {
        let ub: f32 = n as f32 * logf32(n as f32 * logf32(n as f32));
        get_primes_upto(ceilf32(ub) as usize)[n - 1]
    }
}

File: c/libs/prime.c
#include "prime.h"
#include "structs.h"
#include <math.h>
#include <stdint.h>
#include <stdlib.h>

struct SieveAndPrimes main_helper(unsigned int N) {
  int8_t *sieve = (int8_t *)calloc(sizeof(int8_t), N + 1);
  vec primes;
  new_vec(&primes);
  for (unsigned int x = 2; x <= sqrt(N); x++) {
    if (!sieve[x]) {
      for (unsigned int y = x * x; y <= N; y += x)
        sieve[y] = 1;
      vec_push_back(&primes, x);
    }
  }
  struct SieveAndPrimes ret = {primes, sieve};
  return ret;
};

struct SieveAndPrimes get_sieve_and_primes(unsigned int N) {
  struct SieveAndPrimes a = main_helper(N);
  for (unsigned int x = ceil(sqrt(N)); x <= N; x++) {
    if (!a.sieve[x]) {
      vec_push_back(&a.primes, x);
    }
  }
  return a;
}

int8_t *make_sieve(unsigned int N) {
  struct SieveAndPrimes a = main_helper(N);
  vec_free(&a.primes);
  return a.sieve;
}

vec get_primes_upto(unsigned int N) {
  struct SieveAndPrimes a = get_sieve_and_primes(N);
  free(a.sieve);
  return a.primes;
}

unsigned int get_nth_prime(unsigned int n) {
  double ub = n * log(n * log(n));
  return get_primes_upto(ceil(ub)).items[n - 1];
}
